File: platform/logging-service/src/LoggingService.Core/Models/ValidationResult.cs
namespace LoggingService.Core.Models;

/// <summary>
/// Result of a validation operation.
/// </summary>
public sealed record ValidationResult
{
    /// <summary>
    /// Indicates whether the validation passed.
    /// </summary>
    public bool IsValid { get; init; }

    /// <summary>
    /// List of validation errors if validation failed.
    /// </summary>
    public IReadOnlyList<FieldError> Errors { get; init; } = [];

    /// <summary>
    /// Creates a successful validation result.
    /// </summary>
    public static ValidationResult Valid() => new() { IsValid = true };

    /// <summary>
    /// Creates a failed validation result with errors.
    /// </summary>
    public static ValidationResult Invalid(IEnumerable<FieldError> errors) =>
        new() { IsValid = false, Errors = errors.ToList() };
}

/// <summary>
/// Represents a validation error for a specific field.
/// </summary>
public sealed record FieldError
{
    /// <summary>
    /// Name of the field that failed validation.
    /// </summary>
    public required string Field { get; init; }

    /// <summary>
    /// Error code for programmatic handling.
    /// </summary>
    public required string Code { get; init; }

    /// <summary>
    /// Human-readable error message.
    /// </summary>
    public required string Message { get; init; }

    public FieldError() { }

    public FieldError(string field, string code, string message)
    {
        Field = field;
        Code = code;
        Message = message;
    }
}

File: services/user-service/src/main/java/com/authplatform/usersvc/shared/validation/ValidationResult.java
package com.authplatform.usersvc.shared.validation;

import java.util.List;

/**
 * Result of validation operation.
 */
public record ValidationResult(boolean valid, List<FieldError> errors) {
    
    public static ValidationResult success() {
        return new ValidationResult(true, List.of());
    }
    
    public static ValidationResult failure(List<FieldError> errors) {
        return new ValidationResult(false, errors);
    }
    
    public static ValidationResult failure(FieldError error) {
        return new ValidationResult(false, List.of(error));
    }
}
